### PR TITLE
Additional effort to distil things further by separating the algorithmic logic from secondary operations

### PR DESCRIPTION
**Pull request by kirk86**
Hi folks, this pull request introduces a couple of new things.
1) It introduces an api consistent for rapid prototyping in a research oriented environment. The structure of the API is the following:

                                                                        -
                                                                        +-------------------------------+
                                                                        |                               |
                                                                        |            TfUtil             |
                                                                        |                               |            
                                                                        +-------------------------------+
                                                                        |                               |
                                                                        |  Exposes primitive tensorflow |
                                                                        |  operations                   |
                                                                        |                               |
                                                                        |                               |
                                                                        +--------------^----------------+
                                                                                       |
                                                                                       |
                                                                        +--------------+----------------+
                                                                        |                               |
                                                                        |                               |
                                                                        |         Model                 |
                                                                        |                               |
                                                                        +-------------------------------+
                                                                        |                               |
                                                                        |   Exposes primitive model     |
                                                                        |   operations to derived models|
                                                                        |                               |
                                                                        +----------^--------------------+
                                                                                   |
                                               +-----------------------------------+-------------------\
                                               |                                                        --------------------+
                                               |                                                                            |
                                               |                                                                            |
                                               |                                                                            |
                                               |                                                                            |
                                               |                                                                            |
                                  +------------+------------+                                                               |
                                  |                         |                                                    +----------+-------------+
                                  |       Algorithm 1                                                            |   Algorithm 2          |
                                  |                         |                                                    |                        |
                                  +-------------------------+                                                    +------------------------+
                                  |                         |                                                    |                        |
                                  |                         |                                                    |                        |
                                  |                         |                                                    |                        |
                                  +--------------^----------+                                                    +-----------^------------+
                                                 +--------------------------------+------------------------------------------+
                                                                                  |
                                                                                  |
                                                                                  |
                                                                                  |
                                                                                  |
                                                                                  |
                                                                      +-----------+---------------+
                                                                      |                           |
                                                                      |    Policies               |
                                                                      |                           |
                                                                      +---------------------------+
                                                                      |                           |
                                                                      |                           |
                                                                      |                           |
                                                                      |                           |
                                                                      |                           |
                                                                      |                           |
                                                                      +---------------------------+

Also one once has derived the "Model" can easily use tensorflow native operations switch between tf.nn. and tf.layers. whenever they wish to.

2) Separation of logic of the algorithm form the rest of the code. This will makes things more transparent and clear and at the same time will reduce the overhead and entry barrier for folks who might already know stuff about ml but haven't done much of RL.

3) The separation of step 2 requires introducing a new folder structure for reorganizing things in a way to help with maintainability and cleaner code.

image

Personally I would like to see the double `baselines/baselines` directory removed and move everything in the top directory: 
```
---baselines
        +algos
        +utils
         .
         .
        etc.
```
I think also the naming convention is pretty self explanatory. The only thing that one might stumble upon is the `dstruct` folder which stands for data structures. Experience replay, memory, buffers etc. go there. The examples folder now contains all the run_whatever was before only that now are renamed into run_algorithmxyz. In those files I have tried to port the session and training loops, so that the actual algorithms and policies are clean of noise. Also I've tried to separate things according to their functionality, for instance an algorithm/model should have methods such as loss, optimize, predict, etc.

Not really helpful to see everything being cranked up in the constructor. Doesn't make things easy to maintain and additionally to understand, at least that's the case for myself, maybe I'm just too stupid.

4) I've ported the following algorithms so far a2c, acer, acktr, ppo1. 

5) I've tried to maintain PEP and flake8 consistency.

ToDo-list:
1) port the rest
2) clean any sessions being thrown back and forth between objects and put them in a central place like in the examples folder 
3) Create primitive policies which allow to easily combine and create abstract and complicated architectures as needed.

### CODE REVIEW

**Comment by pzhokhov** [comment]
Hi @kirk86 ! Sorry about the review delay and thanks for the great cleanup! We have undertaken similar steps too (as you can see by the score of merge conflicts); but if you are still interested in contributing your work, let's figure out the right steps. Minor note - ppo1 functionality is (mostly) being implemented in ppo2, so we are sunsetting ppo1; I don't think there is any need to refactor that. 
Removal of baselines/baselines - is there a way to do that and still keep "from baselines.ppo2 import ..." pattern? To my knowledge, project_name/pkg_name/subfolder/submodule.py is a standard way of packaging python pacakges; in our case baselines happens to be both project_name and pkg_name.

**Comment by kirk86** [comment]
Hi @pzhokhov!
Thanks for getting back to me.
> Hi @kirk86 ! Sorry about the review delay and thanks for the great cleanup! We have undertaken similar steps too (as you can see by the score of merge conflicts);

Indeed I've noticed that as well 😉 , nice work! 

> but if you are still interested in contributing your work, let's figure out the right steps.

Definitely I'm still interested and I'm more than happy to help and contribute back!

> Minor note - ppo1 functionality is (mostly) being implemented in ppo2, so we are sunsetting ppo1; I don't think there is any need to refactor that.

Nice, thanks for letting me know about it 😃 

> Removal of baselines/baselines - is there a way to do that and still keep "from baselines.ppo2 import ..." pattern? To my knowledge, project_name/pkg_name/subfolder/submodule.py is a standard way of packaging python pacakges; in our case baselines happens to be both project_name and pkg_name.

Yup, your last sentence basically sums everything up. Maybe renaming things into `baselines.rl.subfolder.submodule`? would be more descriptive and later allowing for additional stuff to be added on baselines, e.g. `baselines.pgm.subfolder.submodule`
